opentelemetry-exporter-geneva/geneva-uploader/src/ingestion_service/mod.rs

```diff
@@ -0,0 +1,240 @@
+    /// export GENEVA_PARALLEL_UPLOADS="10"
+    ///
+    /// cargo test test_parallel_uploads -- --ignored --nocapture
+    /// Output:
+    // 🔥 Performing warm-up upload...
+    // 🔥 Warm-up upload complete in 222.42ms
+    // 🚀 Launching 5 parallel uploads...
+    // ✅ Upload 2 complete in 120.43ms. Ticket: ...
+    // ✅ Upload 4 complete in 120.35ms. Ticket: ...
+    // ✅ Upload 3 complete in 120.50ms. Ticket: ...
+    // ✅ Upload 1 complete in 154.62ms. Ticket: ...
+    // ✅ Upload 0 complete in 154.65ms. Ticket: ...
+    // 📊 Average upload duration: 133.60 ms
+    // ⏱️ Total elapsed for 5 parallel uploads: 154.93ms
+
+    #[tokio::test(flavor = "multi_thread")]
+    #[ignore]
+    async fn test_parallel_uploads() {
+        use std::env;
+        use std::time::Instant;
+
+        // Read parallelism level from env
+        // Use env variable if provided, else saturate all tokio threads by default (num_cpus::get())
+        let parallel_uploads: usize = env::var("GENEVA_PARALLEL_UPLOADS")
+            .ok()
+            .and_then(|v| v.parse().ok())
+            .unwrap_or_else(num_cpus::get);
+        let ctx = test_helpers::build_test_upload_context().await;
+
+        // --- Warm-up: do the first upload to populate the token cache ---
+        println!("🔥 Performing warm-up upload...");
+        let start_warmup = Instant::now();
+        let _ = ctx
+            .uploader
+            .upload(ctx.data.clone(), &ctx.event_name, &ctx.event_version)
+            .await
+            .expect("Warm-up upload failed");
+        println!(
+            "🔥 Warm-up upload complete in {:.2?}",
+            start_warmup.elapsed()
+        );
+
+        println!("🚀 Launching {parallel_uploads} parallel uploads...");
+
+        let start_all = Instant::now();
+
+        let mut handles = vec![];
+        for i in 0..parallel_uploads {
+            let uploader = ctx.uploader.clone();
+            let data = ctx.data.clone();
+            let event_name = ctx.event_name.to_string();
+            let event_version = ctx.event_version.to_string();
+
+            let handle = tokio::spawn(async move {
+                let start = Instant::now();
+                let resp = uploader
+                    .upload(data, &event_name, &event_version)
+                    .await
+                    .unwrap_or_else(|_| panic!("Upload {} failed", i));
+                let elapsed = start.elapsed();
+                println!(
+                    "✅ Upload {} complete in {:.2?}. Ticket: {}",
+                    i, elapsed, resp.ticket
+                );
+                elapsed
+            });
+
+            handles.push(handle);
+        }
+
+        let results: Vec<_> = futures::future::join_all(handles)
+            .await
+            .into_iter()
+            .map(|res| res.expect("Join error in upload task"))
+            .collect();
+
+        let total_time = start_all.elapsed();
+
+        let durations: Vec<_> = results;
```

Comment:
nit: Use the variable name as `durations` from the get-go instead of doing an ownership transfer.